src/Disciplina/Educator/SizedMerkleTree.hs

```diff
@@ -0,0 +1,237 @@
+  | SizedMerkleTree !(SizedMerkleNode a)
+  deriving (Eq, Generic)
+
+
+instance Foldable SizedMerkleTree where
+    foldMap _ SizedMerkleEmpty      = mempty
+    foldMap f (SizedMerkleTree n) = foldMap f n
+
+    null SizedMerkleEmpty = True
+    null _                = False
+
+    length SizedMerkleEmpty      = 0
+    length (SizedMerkleTree n) = fromInteger (toInteger (smrSize (mRoot n)))
+
+instance Show a => Show (SizedMerkleTree a) where
+  show SizedMerkleEmpty = "Empty Sized Merkle tree "
+  show (SizedMerkleTree n) = "Sized Merkle tree:  " <> Prelude.show n
+  --show tree = "Merkle tree: "  <> Prelude.show (toList tree)
+
+type LeafIndex = Int
+
+data SizedMerkleNode a
+    = SizedMerkleBranch { mRoot  :: !(SizedMerkleRoot a)
+                        , mLeft  :: !(SizedMerkleNode a)
+                        , mRight :: !(SizedMerkleNode a)}
+    | SizedMerkleLeaf { mRoot  :: !(SizedMerkleRoot a)
+                      , mIndex :: !LeafIndex
+                      , mVal   :: !a}
+    deriving (Eq, Show, Generic)
+
+instance Foldable SizedMerkleNode where
+  foldMap f x = case x of
+    SizedMerkleLeaf {mVal} -> f mVal
+    SizedMerkleBranch {mLeft, mRight} -> foldMap f mLeft `mappend` foldMap f mRight
+
+toLazyByteString :: Builder -> LBS.ByteString
+toLazyByteString = Builder.toLazyByteStringWith (Builder.safeStrategy 1024 4096) mempty
+
+mkLeaf :: HasHash a => (LeafIndex, a) -> SizedMerkleNode a
+mkLeaf (i, a) =
+    SizedMerkleLeaf
+    { mVal  = a
+    , mIndex = i
+    , mRoot = SizedMerkleRoot (unsafeHash a) -- unsafeHash since we need to hash to ByteString
+                              1 -- size of leaf node is 1
+    }
+
+mkBranch :: SizedMerkleNode a -> SizedMerkleNode a -> SizedMerkleNode a
+mkBranch l r =
+    SizedMerkleBranch
+    { mLeft  = l
+    , mRight = r
+    , mRoot  = mkBranchRootHash (mRoot l) (mRoot r)
+    }
+
+mkBranchRootHash :: SizedMerkleRoot a -- ^ left merkle root
+                 -> SizedMerkleRoot a -- ^ right merkle root
+                 -> SizedMerkleRoot a
+mkBranchRootHash (SizedMerkleRoot (AbstractHash hl) sl)
+                 (SizedMerkleRoot (AbstractHash hr) sr)
+  = SizedMerkleRoot
+     (hash $ toLazyByteString $ mconcat
+        [ byteString (one sl)
+        , byteString (one sr)
+        , byteString (convert hl)
+        , byteString (convert hr) ])
+     (sl + sr)
+
+-- | Smart constructor for 'SizedMerkleTree'.
+mkSizedMerkleTree :: (HasHash a, Foldable t) => t a -> SizedMerkleTree a
+mkSizedMerkleTree x = mkSizedMerkleTree' (zip [0,1..] (toList x))
+
+mkSizedMerkleTree' :: HasHash a => [(LeafIndex, a)] -> SizedMerkleTree a
+mkSizedMerkleTree' [] = SizedMerkleEmpty
+mkSizedMerkleTree' ls = SizedMerkleTree (go lsLen ls)
+  where
+    lsLen = Universum.length ls
+    go :: HasHash a => Int -> [(LeafIndex, a)] -> SizedMerkleNode a
+    go _  [x] = mkLeaf x
+    go len xs = mkBranch (go i l) (go (len - i) r)
+      where
+        i = powerOfTwo len
+        (l, r) = splitAt i xs
```

Comment:
Is this method public?

If it is, are there any guarantees that `LeafIndices` will be
1) in order?
2) consequtive?

Is this tree created at once and never changed again?
How do you push new value?

`splitAt n` has `O(n)` time complexity. You actually traverse the list multitude of times, which is not good.